In [1]:
!pip install geocoder
!pip install folium

In [2]:
import geocoder # import geocoder
import pandas as pd
import matplotlib.cm as cm
import matplotlib.colors as colors
import numpy as np
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes
import folium

In [3]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [4]:
import requests
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [5]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [6]:
len(wiki_data), type(wiki_data)

(3, list)

In [7]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [8]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]

In [9]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


In [10]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [11]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [12]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
2                                            Parkwoods  43.753259 -79.329656  
3                                     Victoria Village  43.725882 -79.315572  
4                            Regent Park, Harbourfront  43.654260 -79.360636  
5                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
6          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
160      The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
165                               Church and Wellesley  43.665860 -79.383160  
168  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
178  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [13]:
combined_data.shape

(103, 5)

creating a copy of combined_data dataframe as d3 dataframe

In [14]:
df3 = combined_data.copy()
#fetching only Torronto data
df3 = df3[combined_data.Borough.str.contains("Toronto")]
df3.head()

Postal Code           Borough                                Neighbourhood  \
4          M5A  Downtown Toronto                    Regent Park, Harbourfront   
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
13         M5B  Downtown Toronto                     Garden District, Ryerson   
22         M5C  Downtown Toronto                               St. James Town   
30         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  
4   43.654260 -79.360636  
6   43.662301 -79.389494  
13  43.657162 -79.378937  
22  43.651494 -79.375418  
30  43.676357 -79.293031

In [15]:
toronto_map = folium.Map(location=[43.65, -79.4], zoom_start=12)

X = df3['Latitude']
Y = df3['Longitude']
Z = np.stack((X, Y), axis=1)

kmeans = KMeans(n_clusters=4, random_state=0).fit(Z)

clusters = kmeans.labels_
colors = ['red', 'green', 'blue', 'yellow']
df3['Cluster'] = clusters

for latitude, longitude, borough, cluster in zip(df3['Latitude'], df3['Longitude'], df3['Borough'], df3['Cluster']):
    label = folium.Popup(borough, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=7,
        popup=label,
        color='black',
        fill=True,
        fill_color=colors[cluster],
        fill_opacity=1).add_to(toronto_map)  

toronto_map

# Foursquare Battle


In [16]:
CLIENT_ID = '4L1XNWHFFZ1WFAB41F0HOJLJC4KW4F1AWPPRM3M2OAT204N5' 
CLIENT_SECRET = 'QFSGFOOQMMNB1NQNRRQ5DM3WRTXIMZLUYQ23P5KNISGFU5QS'
VERSION = '20180605' # Foursquare API version

Next, we create a function to get all the venue categories in Toronto

In [17]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood



In [18]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [19]:
venues_in_toronto.shape

(1324, 5)

So we have 1324 records and 5 columns. Checking sample data

In [20]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1          TTC stop #8380               Bus Stop  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

Checking the Venues based on Neighbourhood



In [21]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1311  Mimico NW, The Queensway West, South of Bloor,...   
1312  Mimico NW, The Queensway West, South of Bloor,...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656          TTC stop #8380   
2                  43.753259               -79.329656           Variety Store   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1311               43.628841               -79.520999              Wingporium   
1312               43.628841               -79.520999        South St. Burger   
1313               43.628841               -79.520999               Dollarama   
1314               43.628841               -79.520999          Healthy Planet   
1315               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1                  Bus Stop  
2         Food & Drink Shop  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1311            Wings Joint  
1312           Burger Joint  
1313         Discount Store  
1314        Supplement Shop  
1315                 Bakery  

[412 rows x 5 columns]

So there are 412 records for each neighbourhood.

Checking for the maximum venue categories



In [22]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wine Shop                                Dufferin, Dovercourt Village   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Scarborough Village   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Wine Bar                         43.653206               -79.379817   
Wine Shop                        43.669005               -79.442259   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.744734               -79.239476   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Wine Bar                          The National Club  
Wine Shop                                    Macedo  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[238 rows x 4 columns]

# One Hot encoding the venue Categories

In [23]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1319                  0        0                   0             0   
1320                  0        0                   0             0   
1321                  0        0                   0             0   
1322                  0        0                   0             0   
1323                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1319               0                0                 0                    0   
1320               0                0                 0                    0   
1321               0                0                 0                    0   
1322               0                0                 0                    0   
1323               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Vegetarian / Vegan Restaurant  \
0                0         0  ...                              0   
1                0         0  ...                              0   
2                0         0  ...                              0   
3                0         0  ...                              0   
4                0         0  ...                              0   
...            ...       ...  ...                            ...   
1319             0         0  ...                              0   
1320             0         0  ...                              0   
1321             0         0  ...                              0   
1322             0         0  ...                              0   
1323             0         0  ...                              0   

      Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                    0            0                      0                0   
1                    0            0                      0                0   
2                    0            0                      0                0   
3                    0            0                      0                0   
4                    0            0                      0                0   
...                ...          ...                    ...              ...   
1319                 0            0                      0                0   
1320                 0            0                      0                0   
1321                 0            0                      0                0   
1322                 0            0                      0                0   
1323                 0            0                      0                0   

      Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0            0          0            0              0            0  
1            0          0            0              0            0  
2            0          0            0              0            0  
3            0          0            0              0            0  
4            0          0            0              0            0  
...        ...        ..

In [24]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Vegetarian / Vegan Restaurant  \
0                    0             0  ...                              0   
1                    0             0  ...                              0   
2                    0             0  ...                              0   
3                    0             0  ...                              0   
4                    0             0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 239 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood



In [25]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0             0.000000           0.0  ...   
1               0.0             0.000000           0.0  ...   
2               0.0             0.000000           0.0  ...   
3               0.0             0.000000           0.0  ...   
4               0.0             0.041667           0.0  ...   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                    0.0              0.0       0.0        0.0          0.0   
1                    0.0              0.0       0.0        0.0          0.0   
2                    0.0              0.0       0.0        0.0          0.0   
3                    0.0              0.0       0.0        0.0          0.0   
4                    0.0              0.0       0.0        0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 239 columns]

Let's make a function to get the top most common venue categories



In [26]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods



In [27]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East            Hobby Shop   

       2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot               Skating Rink   
1         Athletics & Sports           Coffee Shop                        Pub   
2                       Bank              Pharmacy  Middle Eastern Restaurant   
3         Chinese Restaurant                  Bank        Japanese Restaurant   
4             Sandwich Place           Coffee Shop         Italian Restaurant   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Dessert Shop  Ethiopian Restaurant     Electronics Store   
1                      Gym           Golf Course     Electronics Store   
2                    Diner         Shopping Mall        Sandwich Place   
3         Department Store  Ethiopian Restaurant     Electronics Store   
4  Comfort Food Restaurant             Juice Bar               Butcher   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant             Drugstore             Donut Shop  
1                    Drugstore            Donut Shop                Dog Run  
2                  Supermarket         Deli / Bodega       Sushi Restaurant  
3  Eastern European Restaurant             Drugstore             Donut Shop  
4          Japanese Restaurant      Sushi Restaurant                   Café

Let's make the model to cluster our Neighbourhood

In [28]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [29]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 4, 0,
       1, 0, 0, 0, 0, 0, 1], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories



In [30]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [31]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
2         M3A        North York                                    Parkwoods   
3         M4A        North York                             Victoria Village   
4         M5A  Downtown Toronto                    Regent Park, Harbourfront   
5         M6A        North York             Lawrence Manor, Lawrence Heights   
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2  43.753259 -79.329656             1.0                  Park   
3  43.725882 -79.315572             0.0           Pizza Place   
4  43.654260 -79.360636             0.0           Coffee Shop   
5  43.718518 -79.464763             0.0        Clothing Store   
6  43.662301 -79.389494             0.0           Coffee Shop   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
2       Food & Drink Shop               Bus Stop           Yoga Studio   
3             Coffee Shop  Portuguese Restaurant          Hockey Arena   
4                    Park                 Bakery               Theater   
5  Furniture / Home Store      Accessories Store   Arts & Crafts Store   
6        Sushi Restaurant                    Gym            Hobby Shop   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
2      Department Store     Electronics Store  Eastern European Restaurant   
3   Distribution Center      Department Store                 Dessert Shop   
4        Breakfast Spot  Gym / Fitness Center                Historic Site   
5           Event Space    Miscellaneous Shop                  Coffee Shop   
6              Beer Bar         Smoothie Shop                Burrito Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2             Drugstore            Donut Shop                Dog Run  
3    Dim Sum Restaurant                 Diner         Discount Store  
4     French Restaurant        Farmers Market            Event Space  
5              Boutique         Women's Store  Vietnamese Restaurant  
6        Sandwich Place                  Café  Portuguese Restaurant

Drop all the NaN values to prevent data skew

In [34]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])


In [35]:
import matplotlib.cm as cm
import matplotlib.colors as colors
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters